mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala

```diff
@@ -462,31 +462,44 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
     val expElogbetaBc = batch.sparkContext.broadcast(expElogbeta)
     val alpha = this.alpha.asBreeze
     val gammaShape = this.gammaShape
+    val logphatPartOptionBase = if (optimizeDocConcentration) Some(BDV.zeros[Double](k)) else None
 
-    val stats: RDD[(BDM[Double], List[BDV[Double]])] = batch.mapPartitions { docs =>
+    val stats: RDD[(BDM[Double], Option[BDV[Double]])] = batch.mapPartitions { docs =>
       val nonEmptyDocs = docs.filter(_._2.numNonzeros > 0)
 
       val stat = BDM.zeros[Double](k, vocabSize)
-      var gammaPart = List[BDV[Double]]()
+      val logphatPartOption = logphatPartOptionBase
       nonEmptyDocs.foreach { case (_, termCounts: Vector) =>
         val (gammad, sstats, ids) = OnlineLDAOptimizer.variationalTopicInference(
           termCounts, expElogbetaBc.value, alpha, gammaShape, k)
-        stat(::, ids) := stat(::, ids).toDenseMatrix + sstats
-        gammaPart = gammad :: gammaPart
+        stat(::, ids) := stat(::, ids) + sstats
+        logphatPartOption.foreach(_ += LDAUtils.dirichletExpectation(gammad))
       }
-      Iterator((stat, gammaPart))
-    }.persist(StorageLevel.MEMORY_AND_DISK)
-    val statsSum: BDM[Double] = stats.map(_._1).treeAggregate(BDM.zeros[Double](k, vocabSize))(
-      _ += _, _ += _)
-    val gammat: BDM[Double] = breeze.linalg.DenseMatrix.vertcat(
-      stats.map(_._2).flatMap(list => list).collect().map(_.toDenseMatrix): _*)
-    stats.unpersist()
-    expElogbetaBc.destroy(false)
-    val batchResult = statsSum *:* expElogbeta.t
+      Iterator((stat, logphatPartOption))
+    }
+
+    val elementWiseSumInPlace = (u : (BDM[Double], Option[BDV[Double]]),
+                                 v : (BDM[Double], Option[BDV[Double]])) => {
+      u._1 += v._1
+      u._2.foreach(_ += v._2.get)
+      u
+    }
+
+    val (statsSum: BDM[Double], logphatOption: Option[BDV[Double]]) = stats
+      .treeAggregate((BDM.zeros[Double](k, vocabSize), logphatPartOptionBase))(
+        elementWiseSumInPlace, elementWiseSumInPlace
+      )
 
+    val batchResult = statsSum *:* expElogbeta.t
     // Note that this is an optimization to avoid batch.count
-    updateLambda(batchResult, (miniBatchFraction * corpusSize).ceil.toInt)
-    if (optimizeDocConcentration) updateAlpha(gammat)
+    val batchSize = (miniBatchFraction * corpusSize).ceil.toInt
+    updateLambda(batchResult, batchSize)
+
+    logphatOption.foreach(_ /= batchSize.toDouble)
+    logphatOption.foreach(updateAlpha(_, batchSize))
+
+    expElogbetaBc.destroy(false)
+    stats.unpersist()
```

Comment:
This line is useless.

Reply:
Do you mean `stats.unpersist()`? Sure, I got rid of it.